UpComingMovies/app/src/main/java/com/example/douglas/upcomingmovies/MainActivity.java

```diff
@@ -0,0 +1,221 @@
+         * user-friendly.
+         */
+        mRecyclerView = (RecyclerView) findViewById(R.id.rv_movieservice);
+        mErrorMessageDisplay = (TextView) findViewById(R.id.tv_error_message_display);
+
+        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
+        mRecyclerView.setLayoutManager(layoutManager);
+
+        mRecyclerView.setHasFixedSize(true);
+
+        mMovieServiceAdapter = new MovieServiceAdapter();
+
+        /**
+         * The number max of pages was locked for avoid overflow of data. This could be
+         * changed for a future release. Therefore this code means that everytime the user
+         * reach(by scrolling) the last movie, the application will do one more request.
+         */
+        mMovieServiceAdapter.setOnBottomReachedListener(new MovieServiceAdapter.OnBottomReachedListener() {
+            @Override
+            public void onBottomReached(int position) {
+                if(lastPageRequested < MAX_PAGES) {
+                    lastPageRequested+=1;
+                    loadMovieData();
+                }
+            }
+        });
+
+        /**
+         * The adapter is the component able to create new items and populates them on the screen
+         * with data.
+         */
+        mRecyclerView.setAdapter(mMovieServiceAdapter);
+
+        loadMovieData();
+    }
+
+    /**
+     * This AsyncTask is almost a must have because the application need to do all the computations
+     * before the draw of the next frame, creating a AsysncTask is a modular aproach leaving the
+     * application free to be rendered while the content is being fetched.
+     */
+    public class UpComingMovieServiceTask extends AsyncTask<URL,Void,List<Movie>>{
+
+        /**
+         * This method is reponsible for call the http request and parse the first part of the
+         * api's response.
+         *
+         * @param urls
+         * @return fetchedMovies
+         */
+        @Override
+        protected List<Movie> doInBackground(URL... urls) {
+            URL searchUrl = urls[0];
+            String movieServiceResults = null;
+
+            try {
+                movieServiceResults = NetworkUtils.getResponseFromHttpUrl(searchUrl);
+                JSONObject jsonResult = new JSONObject(movieServiceResults);
+                JsonHandler jsonHandler = new JsonHandler();
+                JSONArray results = jsonResult.getJSONArray("results");
+                if(fetchedMovies == null){
+                    fetchedMovies = jsonHandler.readResultsArray(results);
+                    getPosterImage(fetchedMovies);
+                }else{
+                    List<Movie> newResult = jsonHandler.readResultsArray(results);
+                    getPosterImage(newResult);
+                    fetchedMovies.addAll(newResult);
+                }
+                return fetchedMovies;
+            } catch (IOException e) {
+                e.printStackTrace();

```

Comment:
What should happen in case of failure? It's a good practice to give users feedback about their actions and its results.

Reply:
Yes, I agree with you, when an Exception was triggered due to resource failures, we should catch the exception and contain the problem, after that we can show some kind of information for the user like "The servers are not responding as usual. Please try again later". When an Exception was triggered due to client code errors, for example, the user is trying to do something that the API does not allow, we(developers) could warn the user about which was the step in the API processing that caused the problem. Therefore a future release of the app would handle exceptions better.